### PR TITLE
Roles

### PR DESCRIPTION
**Pull request by TatsianaMaltseva**
#5

### CODE REVIEW

File: api/Services/IHashPasswords.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿namespace iTechArt.Hotels.Api.Services
+{
+    public interface IHashPasswords
```
Comment (andrei-kruhliakou-itechart) [inline api/Services/IHashPasswords.cs]: What do you need this interface for? Do you really need it?

File: api/Controllers/AuthController.cs [hunk 2]
```diff
@@ -15,40 +18,82 @@ namespace iTechArt.Hotels.Api.Controllers
...
+            }
+            var token = GenerateJWT(account);
             return Ok(token);
         }
 
-        private Account GetAccount(string email, string password) =>
-            _hotelsDb.Accounts.SingleOrDefault(u => u.Email == email && u.Password == password);
+        [Route("registration")]
+        [HttpPost]
+        public async Task<IActionResult> Registrate([FromBody] Login request)
```
Comment (andrei-kruhliakou-itechart) [inline api/Controllers/AuthController.cs]: Register?

**Comment by andrei-kruhliakou-itechart** [comment]
@TatsianaMaltseva You need to merge your new changes with recent changes from main branch first.

**Comment by TatsianaMaltseva** [comment]
> 

I have merged registration branch into roles, and then merged registration into main

File: admin/src/app/Components/authenticationDialog/authenticationDialog.component.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+        <div>
+            <app-login [closeAuthDialog] = "closeAuthDialogGet"></app-login>
+        </div>
+      </mat-tab>
+      <mat-tab label="Register">
+        <div>
+            <app-register [closeAuthDialog] = "closeAuthDialogGet"></app-register>
+        </div>
+      </mat-tab>
+</mat-tab-group>
```
Comment (bagatka) [inline admin/src/app/Components/authenticationDialog/authenticationDialog.component.html]: Something went wrong with indents

File: admin/src/app/Components/authenticationDialog/authenticationDialog.component.html [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<mat-tab-group dynamicHeight>
+      <mat-tab label="Login">
+        <div>
+            <app-login [closeAuthDialog] = "closeAuthDialogGet"></app-login>
```
Comment (bagatka) [inline admin/src/app/Components/authenticationDialog/authenticationDialog.component.html]:
```suggestion
            <app-login [closeAuthDialog]="closeAuthDialogGet"></app-login>
```

File: admin/src/app/Components/authenticationDialog/authenticationDialog.component.html [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+        </div>
+      </mat-tab>
+      <mat-tab label="Register">
+        <div>
+            <app-register [closeAuthDialog] = "closeAuthDialogGet"></app-register>
```
Comment (bagatka) [inline admin/src/app/Components/authenticationDialog/authenticationDialog.component.html]:
```suggestion
            <app-register [closeAuthDialog]="closeAuthDialogGet"></app-register>
```

File: admin/src/app/Components/authenticationDialog/authenticationDialog.component.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+@Component({
+  selector: 'app-authentication-dialog',
+  templateUrl: './authenticationDialog.component.html'
+})
+
+export class AuthenticationDialogComponent {
+  public get closeAuthDialogGet(): Function {
+    return this.closeAuthDialog.bind(this);
+  }
```
Comment (bagatka) [inline admin/src/app/Components/authenticationDialog/authenticationDialog.component.ts:12]: I'm a bit confused why do you need it?

File: admin/src/app/Components/create-admin/create-admin.component.ts [hunk 7]
```diff
@@ -1,50 +1,61 @@
...
   templateUrl: './create-admin.component.html',
   styleUrls: ['./create-admin.component.css']
 })
+
 export class CreateAdminComponent {
-  public registrationForm: FormGroup;
+  public createAdminForm: FormGroup;
+  public hidePassword: boolean = true;
   public serverErrorResponse: string = '';
+  public passwordValidator = new ConfirmValidParentMatcher('notSame');
```
Comment (bagatka) [inline admin/src/app/Components/create-admin/create-admin.component.ts]: The naming is incorrect. It is not a validator

File: admin/src/app/Components/create-admin/create-admin.component.ts [hunk 8]
```diff
@@ -1,50 +1,61 @@
...
     private readonly formBuilder: FormBuilder,
-    private readonly snackBar: MatSnackBar,
-    private readonly accountService: AccountService
+    private readonly snackBar: MatSnackBar
   ) {
-      this.registrationForm = this.formBuilder.group({
+      this.createAdminForm = this.formBuilder.group({
         email: [
-          '',
+          '',
```
Comment (bagatka) [inline admin/src/app/Components/create-admin/create-admin.component.ts]:
```suggestion
          '',
```

Comment (bagatka) [reply admin/src/app/Components/create-admin/create-admin.component.ts]: I would suggest to enable whitespace displaying for your IDE

File: admin/src/app/Components/create-admin/create-admin.component.ts [hunk 9]
```diff
@@ -1,50 +1,61 @@
...
-          '',
+          '', 
           [
             Validators.required,
             Validators.email
           ]
         ],
-        password: ['', [Validators.required]]
-      });
+        password: ['', [Validators.required]],
```
Comment (bagatka) [inline admin/src/app/Components/create-admin/create-admin.component.ts]: Since you have single validator, you can use different approach
```suggestion
        password: ['', Validators.required],
```

File: admin/src/app/Components/create-admin/create-admin.component.ts [hunk 10]
```diff
@@ -1,50 +1,61 @@
...
+        confirmPassword: ['']
+      },
+      { validators: CustomValidators.match('password', 'confirmPassword') }
+      );
```
Comment (bagatka) [inline admin/src/app/Components/create-admin/create-admin.component.ts:55]: Not sure about that indent

File: admin/src/app/Components/create-admin/create-admin.component.ts [hunk 11]
```diff
@@ -1,50 +1,61 @@
...
   }
 
-  public closeRegistrationDialog(): void {
+  public closeCreateAdminDialog(): void {
     this.matDialogRef.close();
   }
-
+
```
Comment (bagatka) [inline admin/src/app/Components/create-admin/create-admin.component.ts]: Why did you add spaces for an empty line? It is called trailing spaces and you should avoid it

File: admin/src/app/Components/login/login.component.html [hunk 12]
```diff
@@ -28,17 +28,17 @@
                 {{ hidePassword ? 'visibility_off' : 'visibility' }}
             </mat-icon>
         </button>
-        <mat-error *ngIf="password?.invalid">
+        <mat-error>
             Password is required
         </mat-error>
     </mat-form-field>
-    <button mat-button (click)="closeAuthDialog()">Back</button>
+    <button mat-button (click)="closeAuthDialog()"> Back </button>
```
Comment (bagatka) [inline admin/src/app/Components/login/login.component.html]:
```suggestion
    <button mat-button (click)="closeAuthDialog()">Back</button>
```

File: admin/src/app/Components/login/login.component.ts [hunk 13]
```diff
@@ -1,20 +1,20 @@
...
+  styleUrls: ['./login.component.css']
 })
-
-export class AuthorizationDialogComponent {
+export class LoginComponent {
   public authForm: FormGroup;
   public showWarning: boolean = false;
   public hidePassword: boolean = true;
 
+  @Input() public closeAuthDialog!: Function;
```
Comment (bagatka) [inline admin/src/app/Components/login/login.component.ts]: Why do you need it?

File: admin/src/app/Components/login/login.component.ts [hunk 14]
```diff
@@ -38,19 +37,16 @@ export class AuthorizationDialogComponent {
...
 
   public login(email: string, password: string): void {
-    this.authService.login(email, password)
+    this.authService
+      .login(email, password)
       .subscribe(
         () =>  { 
           this.closeAuthDialog();
-        }, 
+         },
```
Comment (bagatka) [inline admin/src/app/Components/login/login.component.ts]: Why did you change indent?

File: admin/src/app/Components/navbar/navbar.component.html [hunk 15]
```diff
@@ -1,4 +1,4 @@
 <nav>
     <app-authorization></app-authorization>
-    <app-admin-buttons *ngIf="isLoggedIn"></app-admin-buttons>
+    <app-admin-buttons *ngIf="isLoggedIn && (role === 'admin')"></app-admin-buttons>
```
Comment (bagatka) [inline admin/src/app/Components/navbar/navbar.component.html]: Is it possible to make a method for that?

File: admin/src/app/Components/register/register.component.html [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+        >
+            <mat-icon>
+                {{ hidePassword ? 'visibility_off' : 'visibility' }}
+            </mat-icon>
+        </button>
+        <mat-error>
+            Passwords are not the same
+        </mat-error>
+    </mat-form-field>
+    <button mat-button (click)="closeAuthDialog()"> Back</button>
```
Comment (bagatka) [inline admin/src/app/Components/register/register.component.html]:
```suggestion
    <button mat-button (click)="closeAuthDialog()">Back</button>
```

File: admin/src/app/Components/register/register.component.ts [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+@Component({
+  selector: 'app-register',
+  templateUrl: './register.component.html',
+  styleUrls: ['./register.component.css']
+})
+export class RegisterComponent {
+  public registerForm: FormGroup;
+  public hidePassword: boolean = true;
+  public serverErrorResponse: string = '';
+  public passwordValidator = new ConfirmValidParentMatcher('notSame');
```
Comment (bagatka) [inline admin/src/app/Components/register/register.component.ts]: The same as already mentioned

File: admin/src/app/Components/register/register.component.ts [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+  @Input() public closeAuthDialog!: Function;
```
Comment (bagatka) [inline admin/src/app/Components/register/register.component.ts]: Why?

File: admin/src/app/Components/register/register.component.ts [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+        [
+          Validators.required,
+          Validators.email
+        ]
+      ],
+      password: ['', [Validators.required]],
+      confirmPassword: ['']
+    },
+    { validators: CustomValidators.match('password', 'confirmPassword') }
+    );
```
Comment (bagatka) [inline admin/src/app/Components/register/register.component.ts]: I would suggest to think about better code style for `group` method arguments

File: admin/src/app/Components/register/register.component.ts [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  public register(email: string, password: string): void {
+    this.authService
+      .register(email, password)
+      .subscribe(
+        () => {
+          this.serverErrorResponse = '';
+          this.authService.login(email, password).subscribe();
```
Comment (bagatka) [inline admin/src/app/Components/register/register.component.ts]: Hey @andrei-kruhliakou-itechart .
Can we discuss it? Is it fine to make one more backend request to login in as soon as user registration request finished successfully? Is it possible to modify registration endpoint? What do you think?

File: admin/src/app/Core/customValidators.ts [hunk 21]
```diff
@@ -0,0 +1,31 @@
+import { AbstractControl, FormControl, FormGroupDirective, NgForm, ValidationErrors, ValidatorFn } from '@angular/forms';
+import { ErrorStateMatcher } from '@angular/material/core';
+
+
+export function matchValidator(key1: string, key2: string): ValidatorFn {
```
Comment (bagatka) [inline admin/src/app/Core/customValidators.ts]: What is our chosen code style for empty lines between imports and main code part?

File: admin/src/app/Core/customValidators.ts [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+  return (group: AbstractControl): ValidationErrors | null => {
+    const value1 = group.get(key1)?.value as string;  
+    const value2 = group.get(key2)?.value as string;
+    return value1 === value2 ? null : { notSame: true };
+    };
```
Comment (bagatka) [inline admin/src/app/Core/customValidators.ts]:
```suggestion
  return (group: AbstractControl): ValidationErrors | null => {
    const value1 = group.get(key1)?.value as string;  
    const value2 = group.get(key2)?.value as string;
    return value1 === value2 ? null : { notSame: true };
  };
```

File: admin/src/app/Core/customValidators.ts [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+        private readonly errorKey: string
+    ) {
+    }
+
+    public isErrorState(
+        control: FormControl | null, 
+        form: FormGroupDirective | NgForm | null): boolean {
+            return (control?.touched && control.parent?.hasError(this.errorKey)) === true;
+    }
+}
```
Comment (bagatka) [inline admin/src/app/Core/customValidators.ts]: I think you were using 2 spaces instead of 4 for indents

File: admin/src/app/Core/customValidators.ts [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+
+export class CustomValidators {
+    public static match(key1: string, key2: string): ValidatorFn {
+        return matchValidator(key1, key2);
+    }
+}
+
+export class ConfirmValidParentMatcher implements ErrorStateMatcher {
+
```
Comment (bagatka) [inline admin/src/app/Core/customValidators.ts]: Why do you need this line?

File: api/Controllers/AccountsController.cs [hunk 25]
```diff
@@ -12,24 +14,29 @@ namespace iTechArt.Hotels.Api.Controllers
...
                 return BadRequest("User is already registered with this email");
             }
+            byte[] salt = _hashPasswordsService.GenerateSalt();
             var user = new Account
             {
                 Email = request.Email,
-                Password = request.Password
+                Salt = Convert.ToBase64String(salt),
+                Password = _hashPasswordsService.HashPassword(request.Password, salt),
+                Role = "admin"
```
Comment (bagatka) [inline api/Controllers/AccountsController.cs]: Is there any way to avoid the use of inline role string?
You could make a typo, and it would be pretty hard to figure out why something is not working correctly :)

File: api/Controllers/AuthController.cs [hunk 26]
```diff
@@ -15,40 +18,82 @@ namespace iTechArt.Hotels.Api.Controllers
     public class AuthController : Controller
     {
         private readonly IOptions<AuthOptions> _authOptions;
+        private readonly IHashPasswords _hashPasswordsService;
         private readonly HotelsDatabaseContext _hotelsDb;
 
-        public AuthController(IOptions<AuthOptions> authOptions, HotelsDatabaseContext hotelsDb)
+        public AuthController(IOptions<AuthOptions> authOptions, HotelsDatabaseContext hotelsDb, IHashPasswords hashPasswordsService)
```
Comment (bagatka) [inline api/Controllers/AuthController.cs]: Try to fit within 120 characters on one line.

You've already faced the same challenge with constructor arguments in Angular :)

File: api/Controllers/AuthController.cs [hunk 27]
```diff
@@ -15,40 +18,82 @@ namespace iTechArt.Hotels.Api.Controllers
...
         {
-            var user = GetAccount(request.Email, request.Password);
-            if (user == null)
+            Account account = GetAccountByEmail(request.Email);
+            if (account == null)
             {
                 return Unauthorized();
             }
-            var token = GenerateJWT(user);
+            if (!_hashPasswordsService.CheckIfPasswordIsCorrect(account.Password, request.Password, Convert.FromBase64String(account.Salt)))
```
Comment (bagatka) [inline api/Controllers/AuthController.cs]: Try to fit within 120 characters on one line

File: api/Controllers/AuthController.cs [hunk 28]
```diff
@@ -15,40 +18,82 @@ namespace iTechArt.Hotels.Api.Controllers
...
+        public AuthController(IOptions<AuthOptions> authOptions, HotelsDatabaseContext hotelsDb, IHashPasswords hashPasswordsService)
         {
             _authOptions = authOptions;
             _hotelsDb = hotelsDb;
+            _hashPasswordsService = hashPasswordsService;
         }
 
         [Route("login")]
         [HttpPost]
         public IActionResult Login([FromBody]Login request)
```
Comment (bagatka) [inline api/Controllers/AuthController.cs:37]:
```suggestion
        public IActionResult Login([FromBody] Login request)
```

File: api/Controllers/AuthController.cs (hunk 2 above)
Comment (bagatka) [reply api/Controllers/AuthController.cs]: Why endpoint called `registration`, but the argument class is `Login`

File: api/HotelsDatabaseContext.cs [hunk 29]
```diff
@@ -22,7 +22,18 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
...
+                    .IsFixedLength(true);
+
+                entity.Property(e => e.Salt)
+                    .IsRequired()
+                    .HasMaxLength(24)
+                    .IsUnicode(false)
+                    .IsFixedLength(true);
+
+                entity.Property(e => e.Role)
+                    .HasMaxLength(25)
```
Comment (bagatka) [inline api/HotelsDatabaseContext.cs]: Why `25`?

**Review by bagatka — CHANGES_REQUESTED** [review]

File: admin/src/app/Components/authenticationDialog/authenticationDialog.component.ts (hunk 6 above)
Comment (TatsianaMaltseva) [reply admin/src/app/Components/authenticationDialog/authenticationDialog.component.ts:12]: this losts otherwise, and I can not close parent dialog through child element

File: admin/src/app/Core/customValidators.ts (hunk 21 above)
Comment (TatsianaMaltseva) [reply admin/src/app/Core/customValidators.ts]: There should be one line according to chosen style style

File: api/Controllers/AuthController.cs [hunk 30]
```diff
@@ -47,7 +52,7 @@ public IActionResult Login([FromBody]Login request)
 
         [Route("registration")]
         [HttpPost]
-        public async Task<IActionResult> Registrate([FromBody] Login request)
+        public async Task<IActionResult> Register([FromBody] RegisrtationAccountData request)
```
Comment (andrei-kruhliakou-itechart) [inline api/Controllers/AuthController.cs]: Typo in word Regisrtation

File: admin/src/app/Components/authenticationDialog/authenticationDialog.component.html [hunk 31]
```diff
@@ -1,12 +1,12 @@
 <mat-tab-group dynamicHeight>
     <mat-tab label="Login">
         <div>
-            <app-login [closeAuthDialog]="closeAuthDialogGet"></app-login>
+            <app-login (closeDialogEvent)="closeAuthDialog()"></app-login>
```
Comment (bagatka) [inline admin/src/app/Components/authenticationDialog/authenticationDialog.component.html]: Avoid using names that describe one use case of a component. Use something generic

File: admin/src/app/Components/create-admin/create-admin.component.html [hunk 32]
```diff
@@ -41,7 +41,7 @@ <h1>Add new admin</h1>
             matInput 
             formControlName="confirmPassword"
             [type]="hidePassword ? 'password' : 'text'"
-            [errorStateMatcher]="passwordValidator"
+            [errorStateMatcher]="passwordsStateMatcher"
```
Comment (bagatka) [inline admin/src/app/Components/create-admin/create-admin.component.html:44]: Much better!

File: admin/src/app/Components/login/login.component.ts [hunk 33]
```diff
@@ -13,7 +13,7 @@ export class LoginComponent {
   public showWarning: boolean = false;
   public hidePassword: boolean = true;
 
-  @Input() public closeAuthDialog!: Function;
+  @Output() public closeDialogEvent = new EventEmitter();
```
Comment (bagatka) [inline admin/src/app/Components/login/login.component.ts]:
```suggestion
  @Output() public closeDialogEvent = new EventEmitter<void>();
```

File: admin/src/app/Components/login/login.component.ts [hunk 34]
```diff
@@ -39,6 +39,10 @@ export class LoginComponent {
     });
   }
 
+  public closeAuthDialog(): void {
```
Comment (bagatka) [inline admin/src/app/Components/login/login.component.ts]: Don't forget to change the method name too

File: admin/src/app/auth.service.ts [hunk 35]
```diff
@@ -61,10 +61,21 @@ export class AuthService {
...
+    return this.http
+      .post<string>(
+        `${this.apiUrl}api/auth/registration`,
+        { email, password },
+        options
+      )
+      .pipe(
+        tap((token: string) => {
+          localStorage.setItem(ACCESS_TOKEN_KEY, token);
+        })
```
Comment (bagatka) [inline admin/src/app/auth.service.ts:77]: Why did you choose `tap` and how it works?

File: api/Models/RegisrtationAccountData.cs [hunk 36]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace iTechArt.Hotels.Api.Models
+{
+    public class RegisrtationAccountData
+    {
+        [Required]
+        [EmailAddress]
+        public string Email { get; set; }
+        [Required]
+        public string Password { get; set; }
```
Comment (bagatka) [inline api/Models/RegisrtationAccountData.cs]: That makes code more readable, what do you think @andrei-kruhliakou-itechart?
```suggestion
        [Required]
        [EmailAddress]
        public string Email { get; set; }

        [Required]
        public string Password { get; set; }
```

**Review by bagatka — CHANGES_REQUESTED** [review]

File: admin/src/app/auth.service.ts (hunk 35 above)
Comment (TatsianaMaltseva) [reply admin/src/app/auth.service.ts:77]: I should not change returned from previous method value and I should add token to local storage, it can be called side effect; as I understood, it it the situation "tap" might be used

**Review by andrei-kruhliakou-itechart — APPROVED** [review]

File: api/Models/RegisrtationAccountData.cs (hunk 36 above)
Comment (andrei-kruhliakou-itechart) [reply api/Models/RegisrtationAccountData.cs]: @bagatka I wouldn't make it a requirement but yes, I like the suggested style more.

**Review by bagatka — APPROVED** [review]
